Make demand action accept timestamps & write manual timestamps guide (#1079)
membraneframework/membrane-core: guides/useful_concepts/manual_demands.md

# Manual demands

> Elements with pads using manual flow control have two responsibilities:

Reviewer: I'd start with something like

> This guide explains how to use manual demands. Make sure you read the [introduction to flow control](https://hexdocs.pm/membrane_core/06_flow_control.html) first.
>
> Manual demands are a mechanism for manually controlling the speed of processing data in Membrane pipelines. This mechanism is powerful, but complex - use it only when `:auto` flow control is not sufficient for your use case. 

I'd also link to this guide in https://hexdocs.pm/membrane_core/06_flow_control.html and add examples of when to use and when not to use manual demands either there or here.

- **Output pads**: produce the amount of data requested in the
  [`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`)
  callback, or less if the stream ends (by returning an
  [`:end_of_stream`](`t:Membrane.Element.Action.end_of_stream/0`) action).
- **Input pads**: explicitly request data using the
  [`:demand`](`t:Membrane.Element.Action.demand/0`) action — data only arrives
  after demand has been issued.

## Output pads and `handle_demand`

> When downstream requests data, Membrane invokes

Reviewer 2: ```suggestion
When downstream element requests data, Membrane invokes
```

[`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`) on the
element whose output pad with manual flow control is connected to that

> downstream. The callback receives the pad name, the demanded amount, and the

Reviewer 2: ```suggestion
downstream element. The callback receives the pad name, the demanded amount, and the
```

demand unit. The element is expected to produce and send that amount of data,

> or less if the stream ends.

Reviewer 2: Add some note that of course the element does not need to fulfill the whole demand all at once, in a single callback's invocation

> The unit in which `demand_size` is expressed is resolved as follows:

Reviewer: Let's first explain each demand unit

1. If the **output pad** declares `demand_unit: :buffers | :bytes`, that unit
   is used.

> 2. Otherwise, if the linked **input pad** uses `:buffers` or `:bytes`, that

Reviewer 2: ```suggestion
2. Otherwise, if the linked **input pad** declares `:buffers` or `:bytes` :demand_unit, that
```

> 2. Otherwise, if the linked **input pad** declares `:buffers` or `:bytes` :demand_unit, that

Reviewer: ```suggestion
2. Otherwise, if the linked **input pad** declares `:buffers` or `:bytes` `:demand_unit`, that
```

   unit is inherited.
3. Otherwise — when the input pad uses a timestamp demand unit or has auto

>    flow control — the output pad inherits `:buffers`.

Reviewer: mention that the demand unit is automatically converted to what the output expects if it doesn't match what the input uses

So the output pad can explicitly control the unit it receives demand in, but

> timestamp units are not available on output pads.

Reviewer 2: I don't understand this "timestamp units are not available on output pads." part 🤔

### Declarative nature of `handle_demand`

[`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`) always
receives the **total current outstanding demand** from downstream, not a delta.
There is no need to accumulate demand values across multiple callback invocations 

> — each call tells you the full amount still expected.

Reviewer: Mention that the delta is available in the context

### Producing buffers on demand

A source generating random buffers:

```elixir
defmodule MySource do
  use Membrane.Source

  def_output_pad :output,
    flow_control: :manual,
    demand_unit: :buffers,
    accepted_format: _any

  @impl true
  def handle_demand(:output, demand_size, :buffers, _ctx, state) do
    buffers =
      Enum.map(1..demand_size, fn _ ->
        %Membrane.Buffer{payload: :crypto.strong_rand_bytes(256)}
      end)

    {[buffer: {:output, buffers}], state}
  end
end
```

> A source generating random bytes:

Reviewer 2: Shouldn't it be in a different section that `### Producing buffers on demand`?

Or perhaps we should call this section `### Producing data on demand`?

Reviewer: I think the sentence 'a source generating random bytes/buffers' is not precise - each element generates buffers. I'd say 'source handling demand when the unit is `:buffers`'

```elixir
defmodule MyBytesSource do
  use Membrane.Source

  def_output_pad :output,
    flow_control: :manual,
    demand_unit: :bytes,
    accepted_format: _any

  @impl true
  def handle_demand(:output, demand_size, :bytes, _ctx, state) do
    buffer = %Membrane.Buffer{payload: :crypto.strong_rand_bytes(demand_size)}
    {[buffer: {:output, buffer}], state}
  end
end
```

### Redemand

Sometimes producing all demanded buffers at once is not possible or not
desired — for example when the element generates one buffer at a time. In that
case, return the [`:redemand`](`t:Membrane.Element.Action.redemand/0`) action
alongside the buffer. Membrane will then invoke
[`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`) again

> with the demand reduced by the size of what was just sent.

Reviewer 2: I would mention that:
* generally speaking, `:redemand` should be called when we "know" that there is a chance to fulfill some part of the demand
* we shouldn't return `:redemand` from `handle_demand` in Filters

Reviewer 2: How about this one? 
I think that concerning "we shouldn't return :redemand from handle_demand in Filters" - you can add a tiny note at the end of the "## Filters with manual flow control" section

Issuing `demand: {:input, 5}` sets the demand to 5 regardless of how much
demand was already pending.

This matters when a new demand is issued before the previous one is fully
satisfied. For example, if you demand 5, receive 3 buffers, then demand 5
again:

| Event | Demand |
|-------|--------|
| `demand: {:input, 5}` | 5 |
| buffer received | 4 |
| buffer received | 3 |
| buffer received | 2 |
| `demand: {:input, 5}` | 5 ← overwrites 2 |
| buffer received | 4 |
| buffer received | 3 |
| buffer received | 2 |
| buffer received | 1 |
| buffer received | 0 |

Total received: **8**, not 10. The remaining 2 from the first demand were
discarded when the second `demand: {:input, 5}` was issued.

If you need to accumulate demand rather than overwrite it, use the function
form:

```elixir
{[demand: {:input, &(&1 + 5)}], state}
```

### Timestamp demand units

Timestamp demand units let an element request buffers by specifying a
**timestamp threshold** rather than a count or byte size. The element receives
all buffers up to and including the first one whose timestamp meets or exceeds
the demanded value.

Timestamp demand units are only applicable to **input pads with manual flow

> control**. Output pads do not support them. If an input pad uses a timestamp

Reviewer 2: Let's rephrase it slightly so that to emphasize that the output pad connected to an input pad with `timestamps` demand unit receives demand in `:bytes` or `:buffers`, as specified by output pad's demand unit (defaulting to `:buffers`)

Author: Isn't it said in the line below? TBH I don't know if this commend was written before, or after I added this line, so let me know it the way how it is written right now looks good to you

Reviewer 2: I mean that currently it just describes the scenario when output pad's demand unit in unspecified - what if output pad's demand unit is explicitly set as `:bytes` (or `:buffers`)?

demand unit and the linked upstream output pad does not specify a `demand_unit`,
that element will receive [`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`) with demand expressed in `:buffers`.

#### Available variants

- `{:timestamp, :pts}` — uses each buffer's `:pts` field.
- `{:timestamp, :dts}` — uses each buffer's `:dts` field.
- `:timestamp` or `{:timestamp, :dts_or_pts}` — uses `buffer.dts || buffer.pts`.

#### Timestamp requirements

All buffers passing through an input pad with a timestamp demand unit must have
their relevant timestamp field set to a non-`nil` value — Membrane will raise
an error if a buffer is missing its timestamp. Additionally, timestamps must be

> **monotonically non-decreasing**; non-monotonic timestamps will cause a warning.

Reviewer 2: ```suggestion
**monotonically non-decreasing**; non-monotonic timestamps will cause a warning and undefined behaviour.
```

For `{:timestamp, :pts}`, note that PTS can be non-monotonic in streams with
B-frames (see the [timestamps guide](timestamps.md)). Prefer
`{:timestamp, :dts}` or `{:timestamp, :dts_or_pts}` when DTS is available.

#### How timestamp demands work

The demand value is a `t:Membrane.Time.t/0` value. Use `Membrane.Time`
functions to construct it, for example `Membrane.Time.seconds(1)`.

Timestamps are interpreted relative to an **offset** — the timestamp of the
very first buffer ever received on the pad. This means you never need to know
the absolute starting timestamp of the stream; you only work with durations
from that origin.

For example, if the first buffer arrives with `pts: Membrane.Time.seconds(100)`
and you issue `demand: {:input, Membrane.Time.seconds(1)}`, Membrane will
deliver buffers until it reaches the first one with
`pts >= Membrane.Time.seconds(100) + Membrane.Time.seconds(1)`, i.e.
`pts >= Membrane.Time.seconds(101)`.

> When you issue `demand: {:input, t}`, Membrane delivers all buffered buffers

Reviewer 2: What's the point of this "buffered" ? 🤔 
```suggestion
When you issue `demand: {:input, t}`, Membrane delivers all buffers
```

Issuing a demand at or below that value will result in a warning and no buffers
being delivered.

#### Example: a sink consuming one second of data per second

```elixir
defmodule MyTimestampSink do
  use Membrane.Sink

  # For demonstration purposes only.

  def_input_pad :input,
    flow_control: :manual,
    demand_unit: {:timestamp, :pts},
    accepted_format: _any

  @impl true
  def handle_playing(_ctx, state) do
    {[
       start_timer: {:demand_timer, Membrane.Time.seconds(1)},
       demand: {:input, Membrane.Time.seconds(1)}
     ], %{state | next_demand: Membrane.Time.seconds(2)}}
  end

  @impl true
  def handle_tick(:demand_timer, _ctx, state) do
    {[demand: {:input, state.next_demand}],
     %{state | next_demand: state.next_demand + Membrane.Time.seconds(1)}}
  end

  @impl true
  def handle_buffer(:input, buffer, _ctx, state) do
    File.write!("output.bin", buffer.payload, [:append])
    {[], state}
  end
end
```

Each timer tick advances the demanded threshold by one second: `1s`, `2s`,
`3s`, … All values are measured from the PTS of the very first buffer received
on `:input`, so each demand naturally covers the next one-second slice of the
stream.

> ## Filters with manual flow control

Reviewer 2: I think using `:redemand` in filters in also quite "canonical" and it's less intuitive than using it in sources - how about providing an example for it as well?

The following example combines every two input buffers into one output buffer,
so to satisfy a demand of `n` output buffers, the filter needs `n * 2` input
buffers. 

```elixir
defmodule MyFilter do
  use Membrane.Filter

  def_input_pad :input,
    flow_control: :manual,
    demand_unit: :buffers,
    accepted_format: _any

  def_output_pad :output,
    flow_control: :manual,
    demand_unit: :buffers,
    accepted_format: _any

  @impl true
  def handle_init(_ctx, _opts) do
    {[], %{pending: nil}}
  end

  @impl true
  def handle_demand(:output, demand_size, :buffers, _ctx, state) do
    {[demand: {:input, demand_size * 2}], state}
  end

  @impl true
  def handle_buffer(:input, buffer, _ctx, %{pending: nil} = state) do
    {[], %{state | pending: buffer}}
  end

  def handle_buffer(:input, buffer, _ctx, %{pending: first} = state) do
    output = %Membrane.Buffer{
      payload: first.payload <> buffer.payload,
      pts: first.pts,
      dts: first.dts,
      metadata: first.metadata
    }

    {[buffer: {:output, output}], %{state | pending: nil}}
  end
end
```

> > #### Do not use redemand in a filter's `handle_demand` {: .warning}

Reviewer 2: Something is wrong with this `{: .warning`}

Author: WDYM? It works fine when I run `$ mix docs --open`

>
> Returning [`:redemand`](`t:Membrane.Element.Action.redemand/0`) from
> [`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`) in a

> > filter is illegal. [`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`)

Reviewer: Is it? I don't see it mentioned anywhere in the docs, nor prohibited in the code. I think it rather doesn't make sense, but it's not illegal

Author: AFAIK it is illegal, or at least it was made illegal by recent PRs

> should propagate the demand upstream and return. Processing happens in
> [`handle_buffer/4`](`c:Membrane.Element.WithInputPads.handle_buffer/4`) when
> the data actually arrives.

> ## Auto flow control

Reviewer: I'd move this to the beginning or to https://hexdocs.pm/membrane_core/06_flow_control.html and add an example when auto flow control is not sufficient

Author: Hm, I added it here deliberately, to as a reference for somebody how will read this guide, to remind that auto flow control exists

context holds the amount by which demand increased since the previous invocation,
if you need that delta.

### Producing buffers on demand

A source handling demand when the unit is `:buffers`:

```elixir
defmodule MySource do
  use Membrane.Source

  def_output_pad :output,
    flow_control: :manual,
    demand_unit: :buffers,
    accepted_format: _any

  @impl true
  def handle_demand(:output, demand_size, :buffers, _ctx, state) do
    buffers =
      Enum.map(1..demand_size, fn _ ->
        %Membrane.Buffer{payload: :crypto.strong_rand_bytes(256)}
      end)

    {[buffer: {:output, buffers}], state}
  end
end
```

A source handling demand when the unit is `:bytes`:

```elixir
defmodule MyBytesSource do
  use Membrane.Source

  def_output_pad :output,
    flow_control: :manual,
    demand_unit: :bytes,
    accepted_format: _any

  @impl true
  def handle_demand(:output, demand_size, :bytes, _ctx, state) do
    buffer = %Membrane.Buffer{payload: :crypto.strong_rand_bytes(demand_size)}
    {[buffer: {:output, buffer}], state}
  end
end
```

### Redemand

Sometimes producing all demanded buffers at once is not possible or not

> desired — for example when the element generates one buffer at a time. In that

Reviewer 2: "when the element generates one buffer at a time" - it's a little bit obscure to me, do you mean: "the previous element"?

Author: TBH in my mind the element `generates one buffer at a time` is the same as the one that receives demand, you can think about Membrane.Source scenario

Author: I have rephrased the sentence so maybe now it will sound better

```elixir
defmodule MyTimestampSink do
  use Membrane.Sink

  # For demonstration purposes only.

  def_input_pad :input,
    flow_control: :manual,
    demand_unit: {:timestamp, :pts},
    accepted_format: _any

  @impl true
  def handle_playing(_ctx, state) do
    {[
       start_timer: {:demand_timer, Membrane.Time.seconds(1)},
       demand: {:input, Membrane.Time.seconds(1)}
     ], %{state | next_demand: Membrane.Time.seconds(2)}}
  end

  @impl true
  def handle_tick(:demand_timer, _ctx, state) do
    {[demand: {:input, state.next_demand}],
     %{state | next_demand: state.next_demand + Membrane.Time.seconds(1)}}
  end

  @impl true
  def handle_buffer(:input, buffer, _ctx, state) do
    File.write!("output.bin", buffer.payload, [:append])
    {[], state}
  end
end
```

Each timer tick advances the demanded threshold by one second: `1s`, `2s`,
`3s`, … All values are measured from the PTS of the very first buffer received
on `:input`, so each demand naturally covers the next one-second slice of the
stream.

## Filters with manual flow control

The canonical pattern for a filter with both pads using manual flow control is:

> - **[`handle_demand/5`](`c:Membrane.Element.WithOutputPads.handle_demand/5`)** — propagate the demand upstream by returning a [`:demand`](`t:Membrane.Element.Action.demand/0`) action on the input pad.

Reviewer 2: Let's mention that demand might be modified (just as it's multiplied by 2 in the example below)
